### PR TITLE
chore: delete project api tokens when last mapped project is removed

### PR DESCRIPTION
**Pull request by daveleek**
Removes API token when last project it is mapped to is deleted

### CODE REVIEW

**Comment by daveleek** [comment]
Currently lacks unit tests

File: src/lib/features/project/project-service.ts [hunk 1]
```diff
@@ -558,8 +564,23 @@ export default class ProjectService {
...
         );
 
+        const allTokens = await this.apiTokenStore.getAll();
+        const projectTokens = allTokens.filter(
+            (token) =>
+                (token.projects &&
+                    token.projects.length === 1 &&
+                    token.projects[0] === id) ||
+                token.project === id,
+        );
```
Comment (thomasheartman) [inline src/lib/features/project/project-service.ts:574]: This is essentially looking for tokens that only have a single project, and that project has id `id`, right?

Not important, but is there a nicer way to do it? I think even with 24k tokens it should be fine, but how about something like:

```ts
const tokens = await this.apiTokenStore.getTokensForProject(id);
const toDelete = tokens.filter(token => token.project || token.projects.length === 1)
```

Or something like that? 🤷🏼 Not important, though. Oh, I also don't know if we have a method like that, soooo

**Review by thomasheartman — APPROVED** [review]

File: src/lib/features/project/project-service.e2e.test.ts [hunk 2]
```diff
@@ -2488,6 +2491,103 @@ test('deleting a project with archived flags should result in any remaining arch
     expect(flags.find((t) => t.name === flagName)).toBeUndefined();
 });
 
+test('should also delete api tokens that were only bound to deleted project', async () => {
+    const project2 = 'some';
```
[the remark this answers is not in the document]
Comment (daveleek) [reply src/lib/features/project/project-service.e2e.test.ts]: Good point, copied and modified some tests, didn't want clashes but should have remembered to check for any clashes and renamed properly

File: src/lib/features/project/project-service.e2e.test.ts [hunk 3]
```diff
@@ -2488,6 +2491,103 @@ test('deleting a project with archived flags should result in any remaining arch
...
+    });
+
+    await projectService.deleteProject(project2, user, auditUser);
+    const deletedToken = await apiTokenService.getToken(token.secret);
+    expect(deletedToken).toBeUndefined();
+});
+
+test('should not delete project bound api tokens still bound to project', async () => {
+    const project2 = 'token-deleted-project';
+    const project3 = 'token-not-deleted-project';
```
[the remark this answers is not in the document]
Comment (daveleek) [reply src/lib/features/project/project-service.e2e.test.ts]: As above

File: src/lib/features/project/project-service.ts (hunk 1 above)
Comment (daveleek) [reply src/lib/features/project/project-service.ts]: I couldn't find `this.apiTokenStore.getTokensForProject`, I guess we could add this method though

File: src/lib/features/project/createProjectService.ts [hunk 4]
```diff
@@ -109,6 +111,8 @@ export const createProjectService = (
         eventService,
     );
 
+    const apiTokenStore = new ApiTokenStore(db, eventBus, getLogger);
```
[the remark this answers is not in the document]
Comment (daveleek) [reply src/lib/features/project/createProjectService.ts:114]: I'm fine either way on this one. If we want to move more in the direction of talking between services and not stores then perhaps this is the way to go

[the remark this answers is not in the document]
Comment (thomasheartman) [reply src/lib/features/project/createProjectService.ts]: Sure. Just to be clear: I don't know (and don't really have a strong opinion one way or the other 🤷🏼 ). Just came to think of it.

File: src/lib/features/project/project-service.ts (hunk 1 above)
Comment (thomasheartman) [reply src/lib/features/project/project-service.ts]: Yeah, it was just an idea I had (and def not a blocker). I think that method might be useful, but if this is the only place we do that, then ... eh 🤷🏼

File: src/lib/features/project/project-service.e2e.test.ts [hunk 5]
```diff
@@ -2488,6 +2491,103 @@ test('deleting a project with archived flags should result in any remaining arch
...
+        environment: DEFAULT_ENV,
+        project: project,
+    });
+
+    await projectService.deleteProject(project, user, auditUser);
+    const deletedToken = await apiTokenService.getToken(token.secret);
+    expect(deletedToken).toBeUndefined();
+});
+
+test('should not delete project bound api tokens still bound to project', async () => {
```
Comment (thomasheartman) [inline src/lib/features/project/project-service.e2e.test.ts:2519]:
```suggestion
test('should not delete project-bound api tokens still bound to project', async () => {
```

**Review by thomasheartman — APPROVED** [review]

**Pull request merged**